components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,268 @@
+            &header.magic,
+            cUnstructuredSfaMagicNumber.data(),
+            cUnstructuredSfaMagicNumber.size()
+    );
+    archive_writer.write(reinterpret_cast<char const*>(&header), sizeof(header));
+}
+
+auto write_archive_metadata(FileWriter& archive_writer, std::stringstream const& packed_metadata)
+        -> void {
+    archive_writer.write(packed_metadata.str().data(), packed_metadata.str().size());
+}
+
+auto write_archive_file(std::filesystem::path const& file_path, FileWriter& archive_writer)
+        -> void {
+    FileReader reader(file_path.string());
+    std::array<char, cReadBlockSize> read_buffer{};
+    while (true) {
+        size_t num_bytes_read{};
+        ErrorCode const error_code
+                = reader.try_read(read_buffer.data(), cReadBlockSize, num_bytes_read);
+        if (ErrorCode_EndOfFile == error_code) {
+            break;
+        }
+        if (ErrorCode_Success != error_code) {
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+        archive_writer.write(read_buffer.data(), num_bytes_read);
+    }
+}
+
+auto write_archive_files(
+        FileWriter& archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        std::filesystem::path static_archive_file_path
+                = multi_file_archive_path / static_archive_file_name;
+        write_archive_file(static_archive_file_path, archive_writer);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+    for (size_t i = 0; i < next_segment_id; ++i) {
+        std::filesystem::path segment_path = segment_dir_path / std::to_string(i);
+        write_archive_file(segment_path, archive_writer);
+    }
+}
+}  // namespace
+
+auto write_single_file_archive(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    FileWriter archive_writer;
+    std::filesystem::path single_file_archive_path
+            = multi_file_archive_path.string()
+              + std::string(single_file_archive::cUnstructuredSfaExtension);
+
+    if (std::filesystem::exists(single_file_archive_path)) {
+        throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+    }
+
+    archive_writer.open(
+            single_file_archive_path.string(),
+            FileWriter::OpenMode::CREATE_FOR_WRITING
+    );
+
+    auto const packed_metadata = pack_single_file_archive_metadata(
+            multi_file_archive_metadata,
+            multi_file_archive_path,
+            next_segment_id
+    );
+
+    write_archive_header(archive_writer, packed_metadata.str().size());
+    write_archive_metadata(archive_writer, packed_metadata);
+    write_archive_files(archive_writer, multi_file_archive_path, next_segment_id);
```

Comment:
Now when I think about it again, it might be actually a good idea to let the method takes in packed_metadata (so it has access to file_info), and verify that each archive_file being written has its size matching the one in file_info. This would prevent us creating a corrupted archive if somehow the archive_file changes its size (for example, some random process other than CLP modifies the file)

Edit: just realized that the metadata is already packed, so it's not so easy to access. But I still feel it's worthwhile to check the size of files being written.